### PR TITLE
optimize: refactor the redis session store

### LINKED ISSUE #3082: refactor the redis session data structor
## Why you need it?
 Is your feature request related to a problem? Please describe in details
refactor the redis session data structor,string to hash

## How it could be?
A clear and concise description of what you want to happen. You can explain more about input of the feature, and output of it.

### PR DESCRIPTION
**Pull request by lightClouds917**
### Ⅰ. Describe what this PR did
refactor:the redis session store

### Ⅱ. Does this pull request fix one issue?

fixes #3082

### CODE REVIEW

**Comment by lightClouds917** [comment]
The data struct of session:

image

**Comment by lightClouds917** [comment]
PTAL ,THANKS. @ph3636  @a364176773

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java [hunk 1]
```diff
@@ -90,99 +93,251 @@ public static RedisTransactionStoreManager getInstance() {
...
+     * @param globalTransactionDO
+     * @return
+     */
+    private boolean insertGlobalTransactionDO(GlobalTransactionDO globalTransactionDO) {
+        String globalKey = buildGlobalKeyByTransactionId(globalTransactionDO.getTransactionId());
+        try (Jedis jedis = JedisPooledFactory.getJedisInstance()) {
+            Date now = new Date();
+            globalTransactionDO.setGmtCreate(now);
+            globalTransactionDO.setGmtModified(now);
+            String hmset = jedis.hmset(globalKey, BeanUtils.objectToMap(globalTransactionDO));
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:206]: why not use pipeline

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java [hunk 2]
```diff
@@ -90,99 +93,251 @@ public static RedisTransactionStoreManager getInstance() {
...
-            return true;
+            String xid = jedis.hget(globalKey, REDIS_KEY_GLOBAL_XID);
+            if (StringUtils.isEmpty(xid)) {
+                LOGGER.warn("Global transaction is not exist,xid = {}.Maybe has been deleted by another tc server",
+                        globalTransactionDO.getXid());
+                return true;
+            }
+            Long lrem = jedis.lrem(buildGlobalStatus(globalTransactionDO.getStatus()), 0,
+                    globalTransactionDO.getXid());
+            if (lrem > 0) {
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:243]: you don't have to do that

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java [hunk 3]
```diff
@@ -90,99 +93,251 @@ public static RedisTransactionStoreManager getInstance() {
...
-            pipeline.sync();
-            return true;
+            // Defensive watch to prevent other TC server operating concurrently,Fail fast
+            jedis.watch(globalKey);
+            String previousStatus = jedis.hget(globalKey, REDIS_KEY_GLOBAL_STATUS);
+            if (StringUtils.isEmpty(previousStatus)) {
+                throw new StoreException("Global transaction is not exist, update global transaction failed.");
+            }
+            String previousGmtModified = jedis.hget(globalKey, REDIS_KEY_GLOBAL_GMT_MODIFIED);
+            Transaction multi = jedis.multi();
```
Comment (a364176773) [inline server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:278]: I think the pipe can be used instead

Comment (lightClouds917) [reply server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:310]: > I think the pipe can be used instead

这个地方没有使用pipeline,主要考虑，如果多个tc都来操作，都判断全局事务是存在的，然后都去修改，那会不会出问题。而这几个命令，尤其是hmset，map存不存在，它都会成功。pipeline，看起来防不住这种并发修改的场景。

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java [hunk 4]
```diff
@@ -90,99 +93,286 @@ public static RedisTransactionStoreManager getInstance() {
...
-            pipeline.set(keys, json);
-            pipeline.sync();
-            return true;
+            // Defensive watch to prevent other TC server operating concurrently,Fail fast
+            jedis.watch(globalKey);
+            String previousStatus = jedis.hget(globalKey, REDIS_KEY_GLOBAL_STATUS);
+            if (StringUtils.isEmpty(previousStatus)) {
+                throw new StoreException("Global transaction is not exist, update global transaction failed.");
+            }
+            String previousGmtModified = jedis.hget(globalKey, REDIS_KEY_GLOBAL_GMT_MODIFIED);
```
[the remark this answers is not in the document]
Comment (lightClouds917) [reply server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:316]: > using hmget to get REDIS_KEY_GLOBAL_GMT_MODIFIED，REDIS_KEY_GLOBAL_STATUS in one times?

good idea,i will change

**Review by caohdgege — APPROVED** [review]
LGMT

**Comment by lightClouds917** [comment]
Just optimize 4 places all about  pipelines, remove the unused rollback.PTAL @a364176773 @caohdgege

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java [hunk 5]
```diff
@@ -163,22 +149,8 @@ private boolean deleteBranchTransactionDO(BranchTransactionDO branchTransactionD
...
-                    if (lrem > 0) {
-                        jedis.rpush(branchListKey,branchKey);
-                    }
-                    if (del > 0) {
-                        jedis.hmset(branchKey,branchTransactionDOMap);
-                    }
-                }
-            }
-            return false;
+            pipelined.sync();
```
Comment (caohdgege) [inline server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:152]: what will happen if write timeout here?

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java (hunk 3 above)
Comment (a364176773) [reply server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:258]: I think a direct lock + pipeline is better than Multi

**Review by a364176773 — APPROVED** [review]
LGTM

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java [hunk 6]
```diff
@@ -200,213 +334,153 @@ public GlobalSession readSession(String xid) {
...
 
-    private String getGlobalKeyByXid(String xid) {
-        return DEFAULT_REDIS_SEATA_GLOBAL_SESSION_PREFIX + xid;
+    private String buildGlobalKeyByTransactionId(Long transactionId) {
+        return REDIS_SEATA_GLOBAL_PREFIX + transactionId;
     }
-
-    private String getBranchListKeyByXid(String xid) {
-        return DEFAULT_REDIS_SEATA_XID_BRANCHS_PREFIX + xid;
+    private String buildGlobalKeyByTransactionId(String transactionId) {
```
Comment (ph3636) [inline server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:475]: Repeat method

File: server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java [hunk 7]
```diff
@@ -90,100 +94,230 @@ public static RedisTransactionStoreManager getInstance() {
...
-            Pipeline pipeline = jedis.pipelined();
-            if (StringUtils.isEmpty(branch)) {
-                pipeline.lpush(getBranchListKeyByXid(branchTransactionDO.getXid()), key);
+            String previousBranchStatus = jedis.hget(branchKey, REDIS_KEY_BRANCH_STATUS);
+            if (StringUtils.isEmpty(previousBranchStatus)) {
+                throw new StoreException("Branch transaction is not exist, update branch transaction failed.");
             }
-            pipeline.set(key, JSON.toJSONString(branchTransactionDO));
-            pipeline.sync();
+            Map<String,String> map = new HashMap<>(2);
```
Comment (ph3636) [inline server/src/main/java/io/seata/server/storage/redis/store/RedisTransactionStoreManager.java:171]: Can be changed to Map<String,String> map = new HashMap<>(2, 1);

**Review by ph3636 — APPROVED** [review]
LGTM

**Review by wangliang181230 — APPROVED** [review]
LGTM for @ph3636

**Pull request merged**